pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -116,6 +130,472 @@ impl Default for Options {
+
+        assert!(self.known_ongoing.remove(timeline_id));
+
+        if self.known_ongoing.is_empty() {
+            self.latest = None;
+            tracing::info!("gc is now unblocked following timeline deletion");
+        } else {
+            self.latest = Some((ExistingAttempt::ReadFromIndexPart, witnessed));
+            tracing::info!(
+                n = self.known_ongoing.len(),
+                "gc is still blocked for remaining ongoing detaches"
+            );
+        }
+    }
+}
+
+#[derive(Debug)]
+enum ExistingAttempt {
+    /// Informative; there are still non-zero known ongoing timeline ancestor detaches
+    ReadFromIndexPart,
+
+    /// Exclusive lock carried over tenant reset
+    ContinuedOverRestart(TimelineId),
+
+    /// Exclusive while barrier shows the task running
+    Actual(TimelineId, completion::Barrier),
+}
+
+impl PartialEq<Attempt> for &'_ ExistingAttempt {
+    fn eq(&self, other: &Attempt) -> bool {
+        use ExistingAttempt::*;
+        match self {
+            ReadFromIndexPart => false,
+            ContinuedOverRestart(x) if x == &other.timeline_id => true,
+            Actual(x, barrier) if x == &other.timeline_id && other._guard.blocks(barrier) => true,
+            _ => false,
+        }
+    }
+}
+
+impl PartialEq<&'_ Attempt> for &'_ ExistingAttempt {
+    fn eq(&self, other: &&Attempt) -> bool {
+        self == *other
+    }
+}
+
+impl PartialEq<&'_ TimelineId> for &'_ ExistingAttempt {
+    fn eq(&self, other: &&TimelineId) -> bool {
+        use ExistingAttempt::*;
+        match self {
+            ReadFromIndexPart => false,
+            ContinuedOverRestart(x) | Actual(x, _) if x == *other => true,
+            _ => false,
+        }
+    }
+}
+
+impl ExistingAttempt {
+    fn start_new(&mut self, detached: &TimelineId) -> Result<completion::Completion, Error> {
+        use ExistingAttempt::*;
+        match self {
+            ReadFromIndexPart => {}
+            Actual(other, barrier) if barrier.is_ready() => {
+                if other != detached {
+                    tracing::warn!(prev=%other, next=%detached, "switching ongoing detach; this is not expected to happen normally, but doesn't necessarily mean anything catastrophic");
+                }
+            }
+            Actual(other, _) | ContinuedOverRestart(other) => {
+                return Err(Error::OtherTimelineDetachOngoing(*other))
+            }
+        }
+
+        let (guard, attempt) = Self::new(detached);
+        *self = attempt;
+        Ok(guard)
+    }
+
+    fn new(detached: &TimelineId) -> (completion::Completion, Self) {
+        let (guard, barrier) = completion::channel();
+        (guard, ExistingAttempt::Actual(*detached, barrier))
+    }
+}
+
+/// Represents an across tenant reset exclusive single attempt to detach ancestor.
+#[derive(Debug)]
+pub(crate) struct Attempt {
+    timeline_id: TimelineId,
+
+    _guard: completion::Completion,
+    gate_entered: Option<utils::sync::gate::GateGuard>,
+}
```

Comment:
Would be nice to have separate types for the `gate_entered: None` and `gate_entered: Some` cases. So that one is forced to call `before_shutdown` to get the `gate_entered: None` type.

Reply:
I don't think see the benefit. Earlier I had these stateful enums but they were confusing and cost a lot of lines.